src/commands/basemaps-github/create-mapsheet.ts

```diff
@@ -0,0 +1,106 @@
+import { ConfigBundled, ConfigProviderMemory } from '@basemaps/config';
+import { Bounds } from '@basemaps/geo';
+import { fsa } from '@chunkd/fs';
+import { command, option, optional, string } from 'cmd-ts';
+import * as fgb from 'flatgeobuf/lib/mjs/geojson.js';
+import { FeatureCollection, MultiPolygon } from 'geojson';
+
+import { CliInfo } from '../../cli.info.js';
+import { logger } from '../../log.js';
+import { registerCli, verbose } from '../common.js';
+
+interface Output {
+  sheetCode: string;
+  files: string[];
+}
+
+export const CommandCreateMapSheetArgs = {
+  verbose,
+  path: option({
+    type: string,
+    long: 'path',
+    description: 'Path of flatgeobuf, this can be both a local path or s3 location',
+  }),
+  config: option({
+    type: string,
+    long: 'config',
+    description: 'Path of basemaps config json, this can be both a local path or s3 location',
+  }),
+  output: option({
+    type: string,
+    long: 'output',
+    description: 'Output of the mapsheet file',
+  }),
+  include: option({
+    type: optional(string),
+    long: 'include',
+    description: 'Include the layers with the pattern in the layer name.',
+  }),
+  exclude: option({
+    type: optional(string),
+    long: 'exclude',
+    description: 'Exclude the layers with the pattern in the layer name.',
+  }),
+};
+
+export const basemapsCreateMapSheet = command({
+  name: 'bm-create-mapsheet',
+  version: CliInfo.version,
+  description: 'Create a cog mapsheet from provided flatgeobuf',
+  args: CommandCreateMapSheetArgs,
+  async handler(args) {
+    registerCli(this, args);
+    const path = args.path;
+    const config = args.config;
+    const outputPath = args.output;
+
+    const include = args.include ? new RegExp(args.include.toLowerCase(), 'i') : undefined;
+    const exclude = args.exclude ? new RegExp(args.exclude.toLowerCase(), 'i') : undefined;
+
+    logger.info({ path }, 'MapSheet:LoadFgb');
+    const buf = await fsa.read(path);
+    logger.info({ config }, 'MapSheet:LoadConfig');
+    const configJson = await fsa.readJson<ConfigBundled>(config);
+    const men = ConfigProviderMemory.fromJson(configJson);
+
+    const rest = fgb.deserialize(buf) as FeatureCollection;
+
+    const aerial = await men.TileSet.get('ts_aerial');
+    if (aerial == null) throw new Error('Invalid config file.');
+    const layers = aerial.layers.filter(
+      (c) => c[2193] != null && (c.maxZoom == null || c.maxZoom > 19) && (c.minZoom == null || c.minZoom < 32),
```

Comment:
2193 is used more than once. Is there a useful name for what this index represents that we can use for a `const`?

Reply:
This is easier for us to know which projection for the layer inside config. If we want to change this, probably something like `EpsgCode.Nztm2000`.

The type for layer is a `Record<EpsgCode, string>`.

Reply:
this loop is also run twice, it would be better to put this minZoom/maxZoom check into the loop on L77.

I think it would also be nice to use EpsgCode.NZTM2000 instead of 2193 maybe we should use this convention more?

Reply:
I have updated the EpsgCode.

For the loop, I think it much easier to read the logic if we filter layers first, because filtered layer been used by both L77 and L93. So it much be better to filter it first and avoid the complex condition check in both loops.

Reply:
As this is a copy / paste and used very infrequently I'm not too fused, but all these  loops could be rewritten into a single loop.  

All the imagery is loaded into memory https://github.com/linz/argo-tasks/pull/727/files#diff-bf30fd5dd0d7c140c3054a1e0fe58a85c5a4250fa860f81ce7fad23f0ecdc404R64

so we dont need to "getAll" on https://github.com/linz/argo-tasks/pull/727/files#diff-bf30fd5dd0d7c140c3054a1e0fe58a85c5a4250fa860f81ce7fad23f0ecdc404R82

We could drop it all and have a inner loop that looks something like

https://github.com/linz/argo-tasks/pull/727/files#diff-bf30fd5dd0d7c140c3054a1e0fe58a85c5a4250fa860f81ce7fad23f0ecdc404R93

```typescript
for (const layer of aerial.layers) {
 if (layer.minZoom == null || layer.minZoom < 32) continue;
 if (layer.maxZoom == null || layer.maxZoom > 19) continue; 
 
 const nztmImageryId = layer[EpsgCode.NZTM2000];
 if (nztmImageryId == null) continue;
 
 if (exclude && exclude.test(layer.name)) continue;
 if (include && !include.test(layer.name)) continue;
 
 const img = mem.objects.get(nztmImageryId)
 if (img == null) continue;
 


 // Do bounds check and add current files
}


```

Reply:
I have added some unit test for this then refactoring this into simpler logic. I did a compare of the full output mapsheet.json after refactoring which getting the same output, and the unit test aren't broken neither.

Reply:
LGTM, although I'd love to see more use of random/"any" values in the test inputs so that it's obvious which parts are relevant to the assertion. What do you think, @blacha?

Reply:
While it would be nice to have a bunch more tests, this thing is run maybe once every other month, I'm a little hesitant in adding a bunch more work to this PR as its been sitting around a while now too.